src/platform_impl/linux/x11/xdisplay.rs

```diff
@@ -136,6 +143,34 @@ impl XConnection {
     pub fn default_root(&self) -> &xproto::Screen {
         &self.xcb_connection().setup().roots[self.default_screen]
     }
+
+    /// Get the latest timestamp.
+    #[inline]
+    pub fn timestamp(&self) -> u32 {
+        self.timestamp.load(Ordering::Relaxed)
+    }
+
+    /// Set the last witnessed timestamp.
+    #[inline]
+    pub fn set_timestamp(&self, timestamp: u32) {
+        // Store the timestamp in the slot if it's greater than the last one.
+        let mut last_timestamp = self.timestamp.load(Ordering::Relaxed);
+        loop {
+            if timestamp <= last_timestamp {
```

Comment:
I'm not sure this `max()` behaviour is correct. I would expect timestamp wrapping to be possible. Random Google result is https://tronche.com/gui/x/xlib/input/pointer-grabbing.html saying

> A timestamp is a time value, expressed in milliseconds. It typically is the time since the last server reset. Timestamp values wrap around (after about 49.7 days). The server, given its current time is represented by timestamp T, always interprets timestamps from clients by treating half of the timestamp space as being later in time than T. One timestamp value, named CurrentTime, is never generated by the server. This value is reserved for use in requests to represent the current server time.

I'm too tired right now to do the math. Sounds like something with `wrapping_sub` would be needed to tell apart which of two timestamps is newer.

Basically what libxcb does internally to compare cookies, I guess: https://gitlab.freedesktop.org/xorg/lib/libxcb/-/blob/8935793f1f3751a6aa9d78955c7d6236177986de/src/xcbint.h#L66